File: 151/solution.go
package _151

func reverseWords(s string) string {
	sList := []byte(s)

	reverse := func(x, y int) {
		for x < y {
			sList[x], sList[y] = sList[y], sList[x]
			x++
			y--
		}
	}

	n, i, j, l := len(s), 0, 0, 0
	reverse(0, n-1)
	for j < n {
		if sList[j] != ' ' {
			i = j
			if l != 0 {
				sList[l] = ' '
				l++
			}
			for j < n && sList[j] != ' ' {
				sList[l] = sList[j]
				j++
				l++
			}
			reverse(l-j+i, l-1)
		}
		j++
	}
	return string(sList[:l])
}

File: 151/solution.py
from typing import List


class Solution:
    def reverse(self, s: List[str], i: int, j: int):
        while i < j:
            s[i], s[j] = s[j], s[i]
            i += 1
            j -= 1

    def reverseWords(self, s: str) -> str:
        s = list(s)
        s.reverse()
        x, y, l = 0, 0, 0
        while y < len(s):
            if s[y] != " ":
                x = y
                if l != 0:
                    s[l] = " "
                    l += 1
                while y < len(s) and s[y] != " ":
                    s[l] = s[y]
                    y += 1
                    l += 1
                self.reverse(s, l - y + x, l - 1)
            y += 1
        return "".join(s[:l])
